### PR TITLE
feat(stock): added select field for sales invoice creation

### PR DESCRIPTION
**Pull request by MyuddinKhatri**
[Task](https://bloomstack.com/desk#Form/Task/TASK-2020-01912)
- [x] Added Select Field  **"Auto-Create Invoice on Delivery Note submit"**
- [x] And creation of Sales invoice accordingly if delivered or on submit

### CODE REVIEW

File: erpnext/stock/doctype/delivery_note/delivery_note.py [hunk 1]
```diff
@@ -278,6 +282,37 @@ def validate_packed_qty(self):
 		if has_error:
 			raise frappe.ValidationError
 
+	def make_sales_invoice_for_delivery(self):
```
Comment (Alchez) [inline erpnext/stock/doctype/delivery_note/delivery_note.py:287]: Move this function from bloomstack_core (along with the hooks) to ERPNext

File: erpnext/patches/v13_0/set_auto_create_invoice_on_delivery_note.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import frappe
+
+def execute():
+	frappe.reload_doc("stock", "doctype", "accounts_settings")
+
+	create_invoice_on_submit = frappe.db.get_single_value("Accounts Settings", "auto_create_invoice_on_delivery_note_submit")
```
Comment (Alchez) [inline erpnext/patches/v13_0/set_auto_create_invoice_on_delivery_note.py]: Check if the `auto_create_invoice_on_delivery_note_submit` column exists before doing this call. Refer to `frappe.db.has_column` or something like that

**Review by Alchez — CHANGES_REQUESTED** [review]

**Comment by Alchez** [comment]
@MyuddinKhatri please check Travis patches are failing

File: erpnext/stock/doctype/delivery_note/delivery_note.py [hunk 3]
```diff
@@ -195,6 +195,8 @@ def update_current_stock(self):
 
 	def on_submit(self):
 		self.validate_packed_qty()
+		if frappe.db.get_single_value("Accounts Settings", "auto_create_invoice_on_delivery_note") == "Submit":
+			self.make_sales_invoice_for_delivery()
```
Comment (Alchez) [inline erpnext/stock/doctype/delivery_note/delivery_note.py:198]: This happened on `before_submit` in https://github.com/DigiThinkIT/bloomstack_core/pull/417.

Should this happen on `before_submit` instead?

Comment (MyuddinKhatri) [reply erpnext/stock/doctype/delivery_note/delivery_note.py:198]: Okay

**Review by Alchez — COMMENTED** [review]
@MyuddinKhatri, I can see the old field too somehow, can you look at this one?

image

**Comment by MyuddinKhatri** [comment]
> @MyuddinKhatri, I can see the old field too somehow, can you look at this one?
> 
> image

@Alchez  I removed the field from [bloomstack_core](https://github.com/DigiThinkIT/bloomstack_core/pull/417)

File: erpnext/patches/v13_0/set_auto_create_invoice_on_delivery_note.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import frappe
+
+def execute():
+	frappe.reload_doc("accounts", "doctype", "accounts_settings")
+
+	if frappe.get_meta("Accounts Settings").has_field("auto_create_invoice_on_delivery_note_submit"):
```
Comment (Alchez) [inline erpnext/patches/v13_0/set_auto_create_invoice_on_delivery_note.py:6]: Delete the old field (remove the Custom Field record) after the checks are done. The migrate is not removing it.

**Review by Alchez — CHANGES_REQUESTED** [review]

File: erpnext/patches/v13_0/set_auto_create_invoice_on_delivery_note.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+		create_invoice_on_submit = frappe.db.get_single_value("Accounts Settings", "auto_create_invoice_on_delivery_note_submit")
+		if create_invoice_on_submit:
+			frappe.db.set_value("Accounts Settings", None, "auto_create_invoice_on_delivery_note", "Submit")
+			frappe.delete_doc("Custom Field", "Accounts Settings-auto_create_invoice_on_delivery_note_submit")
```
Comment (Alchez) [inline erpnext/patches/v13_0/set_auto_create_invoice_on_delivery_note.py]: Move this outside of the `if` blocks and convert the function to `frappe.delete_doc_if_exists()`

**Review by Alchez — CHANGES_REQUESTED** [review]

**Review by Alchez — APPROVED** [review]

**Pull request merged**